Adding a search bar to the event feed

lancewf opened the pull request:
### 📋 TODO
- [x] bug with changing selection with arrow keys then clicking selection
- [x] Add where the feed is filtered by event type
- [x] Remove wildcard matching for org and server
- [x] Test the grouping right sidebar with the event feed table
- [x] Write unit tests for the static suggestions for the search bar
- [x] Update docs
- [x] rebase after https://github.com/chef/automate/pull/1102 merges

### :nut_and_bolt: Description: What code changed, and why?
Removing the left sidebar Chef server and orgs filters. Combining the Chef server orgs filters with the event type Combobox to create an event feed search bar. This also changes the route to the event feed page from "/event-feed" to "/dashboards/event-feed"

There were a lot of needed tests added also. 

### :chains: Related Resources
completes #692 

### :+1: Definition of Done
* Left sidebar updated to remove the chef server and org filters
* Added event feed search bar with chef server and org filters

### :athletic_shoe: How to Build and Test the Change
1. `build components/automate-ui && start_all_services`
1. Go to https://a2-dev.test/dashboards/event-feed and inspect

### :white_check_mark: Checklist

- [x] Tests added/updated?
- [x] Docs added/updated?

### :camera: Screenshots, if applicable
![Screen Shot 2019-08-01 at 7 41 37 PM](https://user-images.githubusercontent.com/1679247/62340402-6d189d80-b494-11e9-935c-8f8144246052.png)

![Screen Shot 2019-08-08 at 12 14 45 PM](https://user-images.githubusercontent.com/1679247/62731040-21a34980-b9d6-11e9-93d6-cb87306627f6.png)

![Screen Shot 2019-08-08 at 12 17 51 PM](https://user-images.githubusercontent.com/1679247/62731245-8fe80c00-b9d6-11e9-9a9a-730275f2aafa.png)

lancewf commented on components/automate-ui/src/app/page-components/client-runs-search-bar/client-runs-search-bar.component.html:
```diff
@@ -41,7 +41,13 @@
+        </div>
+        <div *ngIf="!hint.icon" >
+          <div>{{hint.title}}</div>
+        </div>
```
Adding icons if they exists in the suggestion

lancewf commented on components/automate-ui/e2e/event-feed-page.e2e-spec.ts:
```diff
@@ -7,22 +7,6 @@ describe('Event Feed: Guitar String Graph Interaction', () => {
     browser.get('/event-feed');
   });
 
-  it('should have a working dropdown filter', () => {
```
The dropdown filter was removed

lancewf commented on components/automate-ui/src/app/page-components/search-bar/search-bar.component.html:
```diff
@@ -25,7 +25,7 @@
     </div>
   </div>
-  <div *ngIf="suggestionsVisible && selectedCategoryType && search_box.value.length > 0"
+  <div *ngIf="suggestionsVisible && selectedCategoryType && (search_box.value.length > 0 || hasStaticSuggestions())"
```
If there are static suggestions, show them after a category is selected.

lancewf commented on components/automate-ui/src/app/page-components/search-bar/search-bar.component.html:
```diff
@@ -51,13 +57,13 @@
       </li>
-      <li *ngFor="let suggestion of visibleCategories.toJS(); let i = index"
-          (click)="handleCategoryItemClick(suggestion, $event)"
+      <li *ngFor="let category of visibleCategories.toJS(); let i = index"
```
This is only renaming 'suggestion' to 'category' for readability.

lancewf commented on components/automate-ui/src/app/page-components/search-bar/search-bar.component.html:
```diff
@@ -41,7 +41,13 @@
+        <div *ngIf="hint.icon">
+          <div class="suggestion-name-with-icon">{{hint.title}}</div>
+          <chef-icon> {{hint.icon}} </chef-icon>
+        </div>
```
If the suggestion has an icon associated to it display it to the left of the label.

lancewf commented on components/automate-ui/src/app/page-components/search-bar/search-bar.component.scss:
```diff
@@ -66,10 +66,25 @@
 
   &.categories {
     width: 232px;
+
```
Shifting the icon to the left of the label.

lancewf commented on components/automate-ui/src/app/page-components/search-bar/search-bar.component.ts:
```diff
@@ -12,30 +12,30 @@ import {
 import { List } from 'immutable';
 import { clamp, compact } from 'lodash';
-import { Chicklet } from '../../types/types';
+import { SearchBarCategoryItem, Chicklet, SuggestionItem } from '../../types/types';
```
The below changes add static/providedValues to the search bar and allows categories to not allow wildcard filters.

lancewf commented on components/automate-ui/src/app/pages/client-runs/client-runs.component.html:
```diff
@@ -19,14 +19,14 @@
         <chef-subheading>Nodes connected to Chef Automate by way of chef-client.</chef-subheading>
 
-        <app-client-runs-search-bar
+        <app-search-bar
```
A lot of renaming.

lancewf commented on components/automate-ui/src/app/pages/client-runs/client-runs.component.ts:
```diff
@@ -54,70 +60,86 @@ export class ClientRunsComponent implements OnInit, OnDestroy {
 
   // The catagories allowed for searching
-  categoryTypes: Chicklet[] = [
+  categoryTypes: SearchBarCategoryItem[] = [
```
Updating the original user of the search bar component to use the new additions.

lancewf commented on components/automate-ui/src/app/pages/event-feed/event-feed.component.html:
```diff
@@ -13,16 +12,23 @@
+            [filters]="searchBarFilters$ | async"
+            (filtersCleared)="onFiltersClear($event)"
+            (filterRemoved)="onFilterRemoved($event)">
+          </app-search-bar-filter-bar>
```
Adding the search bar to the event feed page.

lancewf commented on components/automate-ui/src/app/pages/event-feed/event-feed.component.html:
```diff
@@ -13,16 +12,23 @@
-              [eventTypeCount]="eventTypeCount"
-              (change)="selectEventType($event)"
-              [disabled]="permissionDenied">
-            </app-event-feed-select>
```
removed the dropdown list. :(

lancewf commented on components/automate-ui/src/app/services/event-feed/event-feed.service.ts:
```diff
@@ -18,63 +18,66 @@ import { ChefEvent,
 
-  getEventFeed(filters: EventFeedFilter,
-    sidebarFilter: SidebarFilter): Observable<ChefEventCollection> {
+  getEventFeed(filters: EventFeedFilter): Observable<ChefEventCollection> {
```
Removing sidebar.

lancewf commented on components/automate-ui/src/app/services/event-feed/event-feed.service.ts:
```diff
@@ -269,8 +240,32 @@ export class EventFeedService {
+    }, searchParam, filters);
+  }
+
+  private addChildEventTypes(filters: Chicklet[]): Chicklet[] {
```
A few event types have child event types that should be included in the filter

lancewf commented on components/automate-ui/src/app/page-components/event-feed-table/event-feed-table.component.ts:
```diff
@@ -17,72 +15,63 @@ import { NgrxStateAtom } from '../../ngrx.reducers';
 export class EventFeedTableComponent implements OnDestroy, OnInit {
   @Input() events: ChefEvent[];
+  // The currently set collection of searchbar filters
+  @Input() searchBarFilters: Chicklet[];
```
Used to request left sidebar grouped events

mjingle commented on components/automate-chef-io/content/docs/event-feed.md:
```diff
@@ -36,10 +36,10 @@ In the event feed, events of the same type by the same user are grouped. The eve
 ## Filtering events
 
-To filter the event feed and event timeline by event type use the drop-down menu in the upper right corner. It defaults to show all events. Available event type filters are clients, cookbooks, data bags, environments, nodes, policyfiles, profiles, roles, and scan jobs.
+To filter the event feed and event timeline by event type, Chef Server or Chef Server organization use the search bar. Available event type filters are clients, cookbooks, data bags, environments, nodes, policyfiles, profiles, roles, and scan jobs.
```
```suggestion
To filter the event feed and event timeline by event type, Chef Server, or Chef Organization, use the search bar. Available event type filters are clients, cookbooks, data bags, environments, nodes, policyfiles, profiles, roles, and scan jobs.
```

mjingle commented on components/automate-chef-io/content/docs/event-feed.md:
```diff
@@ -36,10 +36,10 @@ In the event feed, events of the same type by the same user are grouped. The eve
 ![Event feed filters by type](/images/docs/event-feed-filters-types.png)
 
-To filter by Chef Server or Chef Server organization use the filters available in the sidebar on the left hand side. The sidebar shows all Chef Servers and all available Chef Server organizations. You cannot filter compliance events--profiles and scan jobs--by organization or Chef Server. Compliance events are not visible when either of these filters are applied.
+To filter by Chef Server or Chef Server organization select the Chef Server or Chef Organization in the search bar and start typing the name. You cannot filter compliance events--profiles and scan jobs--by organization or Chef Server. Compliance events are not visible when either of these filters are applied.
```
```suggestion
To filter by Chef Server or Chef Organization, select Chef Server or Chef Organization in the search bar and start typing the name. You cannot filter compliance events -- profiles and scan jobs -- by organization or Chef Server. Compliance events are not visible when either of these filters are applied.
```

lancewf commented:
@mjingle I updated that docs with your suggestions. 

@susanev I am looking into the name change.

phiggins commented on components/automate-chef-io/content/docs/event-feed.md:
```diff
@@ -36,10 +36,10 @@ In the event feed, events of the same type by the same user are grouped. The eve
 ![Event feed filters by type](/images/docs/event-feed-filters-types.png)
 
-To filter by Chef Server or Chef Server organization use the filters available in the sidebar on the left hand side. The sidebar shows all Chef Servers and all available Chef Server organizations. You cannot filter compliance events--profiles and scan jobs--by organization or Chef Server. Compliance events are not visible when either of these filters are applied.
+To filter by Chef Server or Chef Organization, select Chef Server or Chef Organization in the search bar and start typing the name. You cannot filter compliance events -- profiles and scan jobs -- by organization or Chef Server. Compliance events are not visible when either of these filters are applied.
```
Now that server and org are not special cases, does it make sense to include that first sentence?

phiggins commented on components/automate-ui/e2e/event-feed-page.e2e-spec.ts:
```diff
@@ -7,22 +7,6 @@ describe('Event Feed: Guitar String Graph Interaction', () => {
     browser.get('/event-feed');
```
Didn't this URL change in this PR? 🤔

phiggins replied:
Oh I see the only test left in this file is pending. Maybe just remove this file then?

lancewf replied:
I removed this file.

phiggins commented on components/automate-ui/src/app/app.module.ts:
```diff
@@ -89,15 +89,15 @@ import { ApplicationsComponent } from './pages/applications/applications.compone
-  ClientRunsSearchFiltersComponent
-} from './page-components/client-runs-search-filters/client-runs-search-filters.component';
+  SearchBarFilterBarComponent
+} from './page-components/search-bar-filter-bar/search-bar-filter-bar.component';
```
This list of imports is/was alphabetized. At one point there was a linter to enforce this but I'm not sure if that's still the case and I haven't checked if you're getting lint failures on this.

lancewf replied:
I am not seeing any failures in buildkite. I updated the imports to have them alphabetized. 

Is this the only file that has its imports alphabetized?

phiggins replied:
If I remember right, this was something we tried to setup but it didn't work for whatever reason. I think that one list was kept alphabetical just out of convention, since it was so long it was nice to keep it easily scannable.

lancewf commented:
@mjingle @phiggins I updated the docs to combine the two paragraphs talking about the search bar. I removed one of the images to look like below. 
![Screen Shot 2019-08-09 at 2 04 28 PM](https://user-images.githubusercontent.com/1679247/62809018-a95c8700-baae-11e9-907a-8b956a8b156f.png)

lancewf commented:
@susanev I updated the 'entity_type' to 'event-type' with the below commit. 
https://github.com/chef/automate/pull/1106/commits/421b4e8ec92823cc10a9a34cc203a4a2050b3388

![Screen Shot 2019-08-09 at 3 28 38 PM](https://user-images.githubusercontent.com/1679247/62812465-a798c080-baba-11e9-951e-8c8620ebf2da.png)
